packages/data/src/components/use-dispatch/use-dispatch-with-map.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * External dependencies
+ */
+import { mapValues } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useRef, useEffect, useLayoutEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the dispatchMap from the latest render. If a store update happens between
+ * render and the effect, this could cause missed/stale updates or
+ * inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+const useIsomorphicLayoutEffect =
+	typeof window !== 'undefined' ? useLayoutEffect : useEffect;
+
+/**
+ * Custom react hook for returning aggregate dispatch actions using the provided
+ * dispatchMap.
+ *
+ * Currently this is an internal api only and is implemented by `withDispatch`
+ *
+ * @param {Function} dispatchMap  Receives the `registry.dispatch` function as
+ *                                the first argument and the `registry` object
+ *                                as the second argument.  Should return an
+ *                                object mapping props to functions.
+ * @return {Object}  An object mapping props to functions created by the passed
+ *                   in dispatchMap.
+ */
+const useDispatchWithMap = ( dispatchMap ) => {
+	const registry = useRegistry();
+	const currentDispatchMap = useRef( dispatchMap );
+
+	useIsomorphicLayoutEffect( () => {
+		currentDispatchMap.current = dispatchMap;
+	} );
+
+	return useMemo( () => {
+		const currentDispatchProps = currentDispatchMap.current(
+			registry.dispatch,
+			registry
+		);
+		return mapValues(
+			currentDispatchProps,
+			( dispatcher, propName ) => {
+				if ( typeof dispatcher !== 'function' ) {
+					// eslint-disable-next-line no-console
+					console.warn(
+						`Property ${ propName } returned from dispatchMap in useDispatchWithMap must be a function.`
+					);
+				}
+				return ( ...args ) => currentDispatchMap
+					.current( registry.dispatch, registry )[ propName ]( ...args );
+			}
+		);
+	}, [ dispatchMap, registry ] );
```

Comment:
So this means this is executed each time `dispatchMap` changes and from the implementation of `withDispatch`, the function passed is an inline one which means this will be regenerated on each call. This I believe defeats the purpose here. 

Also, I don't mind if `withDispatch` is not refactored to use a hook right now but I'm opposed to it either.

Reply:
Yeah, but it's fine because it's wrapped in `pure` so that will only happen when props change.

Reply:
I'd still like someone else to verify my findings with performance tests but from the ones I've run the impact is pretty much nil either way (though I have seen more _small_  improvements in successive runs than regressions).

Reply:
The idea of the proxy dispatches is to avoid running the mapping function each time the props change to avoid rerendering sub-components (have always the same reference to the functions returned by `withDispatch` across rerenders). This was one of the main differences with `connect`.

How is this addressed if the the action creators are regenerated each time props change.

> I'd still like someone else to verify my findings with performance tests but from the ones I've run the impact is pretty much nil either way

I think ideally, you should disable the "async" mode and do the performance tests to compare exactly the impact. The async mode hides some of the impact by only focusing on the selected block which is going to be rerendered anyway.

Reply:
You are right. If we assume that the keys of the returned object won't change. We can do what I described here: https://github.com/WordPress/gutenberg/pull/15896#discussion_r288836476 and pass in `[]` as the deps in `withDispatch`. That would achieve what we want.

Reply:
> If we assume that the keys of the returned object won't change. We can do what I described here: #15896 (comment) and pass in [] as the deps in withDispatch. That would achieve what we want.

As I mentioned in my reply to your previous comment, the problem here is what would we use as the value for deps from `withDispatch`?  We can't use `Object.values( props )` because of the same issue encountered in `useSelect` because the props object shape is not consistent.  

> I think ideally, you should disable the "async" mode and do the performance tests to compare exactly the impact. The async mode hides some of the impact by only focusing on the selected block which is going to be rerendered anyway.

Do you have some pointers on how I can do that for the performance tests?  If you don't have anything in mind I can look into that (I'm assuming just temporarily modify the AsyncMode provider? for the build the tests are run on).  Although I'm not sure that will give accurate results either because then we won't know if the performance hit is from the selects in the tests or the dispatch right?

With the uncertainty around impact here, I'm inclined to just do `useDispatch` in this pull and leave `withDispatch` as is.

Reply:
> Do you have some pointers on how I can do that for the performance tests? If you don't have anything in mind I can look into that (I'm assuming just temporarily modify the AsyncMode provider?

Pass false to this prop https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/index.js#L212

>  Although I'm not sure that will give accurate results either because then we won't know if the performance hit is from the selects in the tests or the dispatch right?

yes, but I was thinking that if you compare this branch with master, the only changes are in `withDispatch`

Reply:
>As I mentioned in my reply to your previous comment, the problem here is what would we use as the value for deps from withDispatch? We can't use Object.values( props ) because of the same issue encountered in useSelect because the props object shape is not consistent.

An empty array, it never needs to re-run because the callbacks reference the mutable ref.
In this case `deps` would be: depencies that make `Object.keys(mapDispatch(...))` return a different list.

Reply:
> yes, but I was thinking that if you compare this branch with master, the only changes are in withDispatch

Ahh right.

>An empty array, it never needs to re-run because the callbacks reference the mutable ref. In this case deps would be: depencies that make Object.keys(mapDispatch(...)) return a different list.

I apologize for not quite getting it yet.  Do you mean `withDispatch` would call `useDispatchWithMap` like this:

```js
const dispatchProps = useDispatchWithMap( dispatchMap, [] );
```

And then the signature for the hook would be something like this:

 ```js
const useDispatchWithMap( dispatchMap, deps ) {
    dispatchMap = useCallback( dispatchMap, deps );
    /** mostly existing logic **/
}
```

Reply:
My bad, I should have been more clear.

Yeah exactly, but the ref should always be updated from the passed in parameter (not the memoized variable).

Now that I think about it more. Replacing `mapDispatch` with `...deps` in the `useMemo` `deps` would be cleaner and have the same effect.